### PR TITLE
Add split image script

### PR DESCRIPTION
**Pull request by karanphil**
Here is a small script that splits an image at certain given indices along the last dimension (so all the directions). I needed this to split a nifti file that had something like this: b-values = 0, 1500, 1500, 1500, 1500, 0, 1500, 1500, 1500, 1500, splited into separate dwi1_b0.nii.gz, dwi2_b0.nii.gz, dwi1_b1500.nii.gz and dwi2_b1500.nii.gz. The command for this particular example would be:

`scil_split_image.py dwi.nii.gz dwi.bval dwi.bvec 1 5 6 --out_dwi dwi1_b0.nii.gz dwi1_b1500.nii.gz dwi2_b0.nii.gz dwi2_b1500.nii.gz --out_bval dwi1_b0.bval dwi1_b1500.bval dwi2_b0.bval dwi2_b1500.bval --out_bvec dwi1_b0.bvec dwi1_b1500.bvec dwi2_b0.bvec dwi2_b1500.bvec`

where the input is dwi.nii.gz, dwi.bval and dwi.bvec.

By default, if no output names are given, the script will take the input names and append the indices to it, such as dwi_0_1.nii.gz, dwi_1_5.nii.gz, dwi_5_6.nii.gz and so on.

This can be easily tested on any data, so I don't think test data is useful for this PR. Also, I added a test file for the script.

### CODE REVIEW

**Comment by karanphil** [comment]
There might be a way to improve the script memory-wise, by not loading the full data at once. Thoughts?

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_split_image.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    if args.out_bvec and len(args.out_bvec) != len(args.split_indices) + 1:
+        parser.error('--out_bvec must contain len(split_indices) + 1 names.')
+
+    bvals, bvecs = read_bvals_bvecs(args.in_bval, args.in_bvec)
+
+    img = nib.load(args.in_dwi)
+    data = img.get_fdata(dtype=np.float32)
+
+    # Check if the indices fit inside the range of possible values
+    if np.max(args.split_indices) > data.shape[-1]:
```
Comment (frheault) [inline scripts/scil_split_image.py:113]: Can you add a check if the indices are in increasing order?
Just in case someone tries to break it?

**Comment by frheault** [comment]
> There might be a way to improve the script memory-wise, by not loading the full data all at once. Thoughts?

If you look at scil_extract_dwi_shell.py, how hard would it be to do that in a memory-friendly way?

File: scripts/scil_split_image.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    add_overwrite_arg(p)
+
+    return p
+
+
+def create_name_from_indices(indices, input):
+    input_path = Path(input)
+    parent_dir = input_path.parent
+    input_name = Path(input_path.stem).stem
+    index_name = "_" + str(indices[0]) + "_" + str(indices[1])
```
Comment (frheault) [inline scripts/scil_split_image.py:77]: Why are you calling this function with a indices[i:i+2]?

Also should indices[1] be [-1] to capture the first and last?

File: scripts/scil_split_image.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+    p.add_argument('in_bvec',
+                   help='The b-vectors file in FSL format (.bvec).')
+
+    p.add_argument('split_indices', nargs='+', type=int,
+                   help='The list of indices where to split the image. For '
+                        'example 3 10. This would split the image in three '
+                        'parts, such as [:3], [3:10], [10:].')
+
+    p.add_argument('--out_dwi', nargs='+', default=[],
```
Comment (frheault) [inline scripts/scil_split_image.py:47]: Would it be better to simply list the basename and the script automatically save nii.gz & bval & bvec?
@arnaudbore what do you think. These extension should be enforced anyway, right?

Comment (karanphil) [reply scripts/scil_split_image.py:47]: You mean give a mandatory out_basename and automatically use my way of saving with the indices? (ex. dwi_0_3.nii.gz)

**Comment by karanphil** [comment]
> > There might be a way to improve the script memory-wise, by not loading the full data all at once. Thoughts?
> 
> If you look at scil_extract_dwi_shell.py, how hard would it be to do that in a memory-friendly way?

I could add a --block-size argument and mimic the way scil_extract_dwi_shell.py does it.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_split_image.py (hunk 3 above)
Comment (frheault) [reply scripts/scil_split_image.py:48]: I mean you list --out_basename dwi_0_0 dwi_0_1500 dwi_1_0 dwi_1_1500 or whatever

Then you know its gonna be .nii.gz .bval and .bvec
Why list 3 times the same thing if you know the extension already

Comment (karanphil) [reply scripts/scil_split_image.py:48]: Ok yes. Should I assume it's .nii.gz, .bval and .bvec, or should I pick the same extensions as the input files? Sometimes, its .nii, .bvecs, .bvals...

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_split_image.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+def main():
+    parser = _build_arg_parser()
+    args = parser.parse_args()
+
+    if args.verbose:
+        logging.basicConfig(level=logging.INFO)
+
+    assert_inputs_exist(parser, [args.in_dwi, args.in_bval, args.in_bvec])
+    assert_outputs_exist(parser, args, [],
+                         optional=args.out_basename)
```
Comment (arnaudbore) [inline scripts/scil_split_image.py:72]: Does not do anything since out_basename is not a file yet.

File: scripts/scil_split_image.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    img = nib.load(args.in_dwi)
+
+    # Check if the indices fit inside the range of possible values
+    if np.max(args.split_indices) > img.shape[-1]:
+        parser.error('split_indices values must be lower than the total '
+                     'number of direcitons.')
+    if np.min(args.split_indices) <= 0:
+        parser.error('split_indices values must be higher than 0.')
+    # Check if the indices are in increasing order
```
Comment (arnaudbore) [inline scripts/scil_split_image.py:84]: Just put a warning and reorder if not in the correct order.

File: scripts/scil_split_image.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        bvals_split = bvals[indices[i]:indices[i+1]]
+        bvecs_split = bvecs[indices[i]:indices[i+1]]
+        # Saving the output files
+        index_name = "_" + str(indices[i]) + "_" + str(indices[i+1])
+        if args.out_basename:
+            data_name = args.out_basename + index_name
+            bval_name = args.out_basename + index_name
+            bvec_name = args.out_basename + index_name
+        else:
+            data_name = str(Path(Path(args.in_dwi).stem).stem) + index_name
```
Comment (arnaudbore) [inline scripts/scil_split_image.py:101]: Use os.path.basename(args.in_dwi).split('.')[0] or out_basename.
It seems more elegant.

**Review by arnaudbore — COMMENTED** [review]
One major issue is still the naming convention.
Once indexes have been validated I would add a function that output the filenames (list of tuples with the indexes used. If indexes are 3 10. You should have 0_3, 4_10, 11_end.

Then you check if these files exist.
And finally you loop you list of tuples with everything needed.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_split_image.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    assert_inputs_exist(parser, [args.in_dwi, args.in_bval, args.in_bvec])
+
+    bvals, bvecs = read_bvals_bvecs(args.in_bval, args.in_bvec)
+
+    img = nib.load(args.in_dwi)
+
+    # Check if the indices fit inside the range of possible values
+    if np.max(args.split_indices) >= img.shape[-1]:
+        parser.error('split_indices values must be lower than the total '
+                     'number of direcitons.')
```
Comment (arnaudbore) [inline scripts/scil_split_image.py:77]:
```suggestion
                     'number of directions.')
```

**Pull request merged**